pkg/storegateway/bucket.go

```diff
@@ -1035,12 +1089,18 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
 			}
 			series.Labels = labelpb.ZLabelsFromPromLabels(lset)
 			if err = srv.Send(storepb.NewSeriesResponse(&series)); err != nil {
+				if cleanup := set.CleanupFunc(); cleanup != nil {
+					cleanup()
+				}
 				err = status.Error(codes.Unknown, errors.Wrap(err, "send series response").Error())
 				return
 			}
+			if cleanup := set.CleanupFunc(); cleanup != nil {
```

Comment:
I'm not sure this is correct. We're calling the cleanup funcs as soon as the 1st series is sent from the batch, but not after all series have been sent. Am I missing anything?

Reply:
the cleanupFunc from the `batchesSeriesSet` is only meant to do anything when invoked on the last series of a batch. Otherwise it should be a noop.


But this reminded me that this is not how it's implemented. because it returns a closure, it will do the clean up on the first cleanUpFunc invocation after loading the next batch. Which is not necessarily the cleanUpFunc invocation on the last series of the previous batch. I will fix this